core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java

```diff
@@ -573,6 +570,51 @@ public void approve(Resource subj, IRI pred, Value obj, Resource ctx) throws Sai
 			addStatement(subj, pred, obj, explicit, ctx);
 		}
 
+		@Override
```

Comment:
Maybe you can just reuse `addStatement` here:

```java
@Override
public void approveAll(Set<Statement> approved, Set<Resource> approvedContexts) {
	sinkStoreAccessLock.lock();
	try {
		for (Statement stmt : approved) {
			addStatement(stmt.getSubject(), stmt.getPredicate(), stmt.getObject(), explicit, stmt.getContext());
		}
	} finally {
		sinkStoreAccessLock.unlock();
	}
}
```
and then modify `addStatement` by checking that the lock is already held by the current thread:

```java
boolean hasLock = sinkStoreAccessLock.isHeldByCurrentThread();
if (! hasLock) {
	sinkStoreAccessLock.lock();
}
try {
	// logic here
} finally {
	if (! hasLock) {
		sinkStoreAccessLock.unlock();
	}
}
```